src/typechecker.c

```diff
@@ -890,7 +889,35 @@ NODISCARD bool typecheck_expression(AST *ast, Node *expr) {
     } break;
 
     /// Make sure a cast is even possible.
-    case NODE_CAST: TODO();
+    case NODE_CAST: {
+      Type *t_to = expr->type;
+      if (type_is_incomplete(t_to))
+        ERR(t_to->source_location, "Can not cast to incomplete type %T", t_to);
+
+      if (!typecheck_expression(ast, expr->cast.value))
+        return false;
+
+      Type *t_from = expr->cast.value->type;
+
+      // TODO: What do we do when from and to types are equal?
+
+      if (type_is_incomplete(t_from))
+        ERR(expr->cast.value->source_location, "Can not cast from an incomplete type %T", t_from);
+
+      // FROM any integer type TO any integer type is ALLOWED
+      if (is_integer(t_from) && is_integer(t_to)) break;
+
+      // FROM any pointer type TO any integer type is ALLOWED
+      if (is_pointer(t_from) && is_integer(t_to)) break;
+
+      // FROM any integer type TO any pointer type is DISALLOWED
+      if (is_integer(t_from) && is_pointer(t_to))
+        ERR(expr->source_location, "Can not cast from integer type %T to pointer type %T", t_from, t_to);
```

Comment:
We might want to allow casting from integers to pointers iff the integer type has the same size as the pointer type. There aren’t many good reasons to do this, but if it’s an explicit cast then I’d say let the user shoot themselves in the foot if they really want to...

Reply:
Oh, and also, some libraries that take function pointers or other pointers often use `((some_pointer_type)1)` etc. as a special value, so there *are* cases where we might want to be able to pass arbitrary integers as pointers, because, in those cases, those ‘pointers’ aren’t even supposed to be dereferenced anyway...

Reply:
In that case I'd say we should just make the external function declaration not use a pointer, as it's an opaque type anyway that is compatible with an architecture-sized integer. I am reluctant to make it so easy to just cast 0 to a pointer and dereference it, lmao.

Reply:
> In that case I'd say we should just make the external function declaration not use a pointer, as it's an opaque type anyway that is compatible with an architecture-sized integer.

Well, the problem is that you’re supposed to pass in either a pointer (or even a function pointer)... or the literal value one. If we make the parameter a number (which is gonna be nigh-on impossible if we ever decide to auto-generate code from C headers), then you have to cast the thing you’re supposed to pass to it (a (function) pointer) to a number just for the one special case to work. If we instead leave it as-is, not only do we not have to change the signature, but we can avoid casting in the general case and require casting in the special case (and actually, the special case would probably just end up being a global constant of the right type that contains the cast in its declaration), so we actually *never* have to cast (at least not as the user).

> I am reluctant to make it so easy to just cast 0 to a pointer and dereference it, lmao.

I mean, I’d say banning casts of numbers to pointers does little to prevent bugs or increase type safety. At least in my experience, 99% of pointer-related bugs and errors are not due to e.g. casting the wrong number to a pointer type (osdev excluded lol). And the cast would have to be explicit too, so if someone *really* wants to write `42 as @integer`, then it’s their fault if it segfaults...

Reply:
If we pass in either a pointer or the value one, could we solve this with an overload? It would require generating overloaded and name mangled wrappers around the actual external symbol

Reply:
You’d have to do this manually for every function in any library that has this behaviour.

Reply:
I mean, I don't think there's going to be any compilers/transpilers targeting this language anytime soon :P
I guess I'm struggling to see why making external function declarations "manual" (or anything for that matter) is really a problem. Unless you have plans that I don't know about :)

Reply:
But wouldn't it just be two external declarations? The overloading is done all by us, right?

Reply:
It’s definitely *possible*, but say you want to use some random C library in a new project. That’s possible in terms of ABI, but you have to manually convert all constants, type declarations, and function declarations that you need.

Manually converting implementation-defined C stdlib internals to another language doesn’t sound like fun.

Reply:
... especially if it’s the Microsoft stdlib...

Reply:
Because of that, doing this automatically by reusing an existing C compiler (like Clang) would probably be easier in the long run; we could also write a C parser ourselves but... do you really want to do that? :eyes: I mean, I suppose that’d be one way to make this thing self-hosted haha

Reply:
I mean, that would be a separate project to the compiler itself, I'd say. Like a conversion script or program `c2un` or whatever it's named :).

> Manually converting implementation-defined C stdlib internals to another language doesn’t sound like fun.

I would agree. 

So let's say `c2un` spits out external C declarations in order to interface with the C library. The point discussed refers to the fact that there's no way this tool could know which pointers need overloads with integers and which ones don't. However, I'd argue that because in C those casts are allowed, then a true interface would just have an overloaded function with `integer` and `pointer` both possible, right? And then those overloaded functions in `un` both just call the actual underlying external symbol, while keeping (safer) type semantics right?

Reply:
> However, I'd argue that because in C those casts are allowed, then a true interface would just have an overloaded function with `integer` and `pointer` both possible, right?

If you do that for every function then you can just pass any integer to it and it would get converted to a pointer; this is worse than explicit casts; what you’re proposing are *implicit* integer-to-pointer casts.

Reply:
And yeah, making `c2un` a separate project would make sense too I’d say

Reply:
Well, they'd only be implicit when they are auto generated explicitly for representing functions written in C, but yeah basically. In this one specific, rare case, the language would allow the looser type system of C to be used when actually interfacing with functions written in C...

Reply:
Yeah, but the thing is: C doesn’t allow that implicitly. Even in C, you have to cast it explicitly. And you’d be able to pass any number to any function accepting any pointer type, which, idk, just doesn’t sound like a good idea to me...

Reply:
> And you’d be able to pass any number to any function accepting any pointer type, which, idk, just doesn’t sound like a good idea to me...

I feel like you keep leaving out the important part

> And you’d be able to pass any number to any function accepting any pointer type |when using `c2un` to automatically convert C headers into `un` code|, which, idk, just doesn’t sound like a good idea to me...

Reply:
Yeah any number to any external C function accepting any pointer type. That’s still not better, or am I missing something again? :eyes:

Reply:
I am not sure if you are missing it, but I'll mention this:

c:
```c
int foo(int *a) {
  return *a;
}

// ... in main ...
int test = 69;
foo(&test);
foo((int *)test);
```

un:
```
foo : ext c integer(a : @integer)
foo : ext c integer(a : integer)

test : integer = 69
foo(&test)
foo(test as @integer)
```

Also I keep forgetting `integer` in `un` and `int` in `c` are not the same, haha.

Reply:
I mean, in that case the overload is redundant since `as @integer` converts it to an `@integer` anyway, so only the first overload is ever gonna be used. Also, this is exactly what I’m proposing :eyes:

Reply:
I see; so I am just tired and bad at explaining myself :kekw:. I should have wrote code sooner, haha. Sorry about that <3

Reply:
Yeah, `as @integer` casts/converts the int to a pointer. That’s what I meant haha.